Lambdify - fast numerical evaluation into Python buffers (no deferred exception)

bjodah opened the pull request:
Here I try out choosing real/complex at instantiation of `Lambdify`.
It is thus an alternative to raising the exception at call as in #3

bjodah commented:
Looks like all failures (except sage download link 404:ed) are related to note not knowning about pytest's xfail and skip.

certik commented:
@bjodah what is the advantage of the xfail and skipped tests? Aren't they just slowing down the test suite for no real reason?

I know there must be a good motivation for it, otherwise the test runners wouldn't implement it, but I always just followed a simple rule that either the test works and then it should be there, or it doesn't work, and then it shouldn't be there.

bjodah commented:
@certik, the skip can be conditional: in this case some tests will be skipped if NumPy is not available (since they rely on it).

The xfail is maybe not quite as useful, I see it as a way of documenting a desired behaviour which one maybe have not had time to implement yet (or if it fails due to a bug in an upstream package which might not be fixed any time soon). It is often at a late stage you realize that you will not be able to make a test pass, so you're faced with a few options: comment it out, throw it away or mark it as xfail.

Anyways, I don't want to be religous about these things and I'll happily comment things out if that is preferred to xfailing. And giving it some thought, the conditional skip may be implemented like this:

```
def test_ravel():
    if HAVE_NUMPY:
        assert np.ravel(some_array) == [1, 2, 3]
    else:
        assert True
```

It would still work with the minor drawback that quite a few tests reported as "Passing" would in reality not have been really run.

So I could comment out xfail tests and refactor skipped ones accordingly if you'd like.

isuruf commented:
That's what I used here, 
https://github.com/symengine/symengine.py/blob/e1136d9016deb434ccea77a4aaae8da7bf000536/symengine/tests/test_sage.py#L5

certik commented:
Can you merge this with the latest master? There are some conflicts.

certik commented:
@bjodah I am fine with various solutions to this problem. I just wanted to understand the motivation. 

A quick, though perhaps a bit dirty solution is as follows:

* You can skip a test by prefixing it with a different name than `test_`. For example you can use `xtest_`. Instead of xfailing, you can just skip it using this method.

* You can conditionally skip it using the technique that @isuruf mentioned above.

That way we can get this PR in and concentrate on actual issues in this PR, rather than our testing framework.

However, in the long run, we should just use standard methods for this in Python. What is wrong with the `sympy.utilities.pytest` solution?

One solution is to just use `py.test`, that's certainly a possibility. We used to only use py.test for sympy, but then there was a serious bug that if you gave two filenames the same name in different places in the directory structure, then one of them would not be run... But I still like py.test the most.

@isuruf any objections against just using py.test? Again, that decision is completely orthogonal to this PR and no matte the outcome, I suggest to just use a quick & dirty solution that I proposed above to get this PR in. But since we started discussing it here, we can as well reach a conclusion here.

certik commented on .travis.yml:
```diff
@@ -102,7 +103,10 @@ install:
+  #  - git checkout 8537c67507fae9b1d8d575d322c6ade4f1a59297
+  # This PR relies proposed modifications of symengine:
+  # see https://github.com/symengine/symengine/pull/623
+  - git remote add bjodah git://github.com/bjodah/symengine.git; git fetch bjodah; git checkout d9c4e3185a798c4aaeaac75363597e40b10517c9
```
Cool, nice way to test things that depend on symengine's PRs that are not yet in.

isuruf commented:
I'm fine with using just `py.test`

certik commented:
Ok, I think we can just use `py.test`, so then @bjodah you don't need to modify your PR, just use py.test.

I am overall OK to the changes here, but we need to make sure all tests pass and perhaps some other technical issues get resolved.

isuruf commented on appveyor.yml:
```diff
@@ -63,7 +63,7 @@ install:
 
 - set PATH=C:\Python%PYTHON_VERSION%;C:\Python%PYTHON_VERSION%\Scripts;%PATH%
-- pip install nose pytest sympy
+- pip install nose pytest sympy numpy
```
I would have just one test with numpy. It takes a very long time to install numpy on appveyor

certik replied:
Good point. Either way though, I think we should install a binary numpy on appveyor.

isuruf commented on bin/install_travis.sh:
```diff
@@ -12,8 +12,13 @@ if [[ "${WITH_SAGE}" != "yes" ]]; then
-
-    conda create -q -n test-environment python="${PYTHON_VERSION}" pip cython sympy nose pytest;
+    CONDA_PKGS="pip cython sympy nose pytest"
+    if [[ "${SKIP_NUMPY}" == "yes" ]]; then
```
I think the option should be `WITH_NUMPY` like other optional dependencies and add `WITH_NUMPY=yes` to a few existing tests.

bjodah replied:
I agree with respect to logic of naming the option, however it would make more sense to do extensive testing of NumPy because this is what we can expect the majority to use. So inverting the case I guess I would add "WITH_NUMPY=yes" to all test cases but one, is that alright?

isuruf replied:
Sure

certik replied:
Yes. We already have tons the `WITH_*` options, so we just add another one. I think it's easier to maintain it that way --- for each test you simply enable what you need.

bjodah commented:
Sorry, but I seem to have messed up the rebasing and introduced duplicate commits.
Any advice on how to safely correct that?

Regarding numpy: It would be nice if we could test it extensively, but if that's not practical with appveyor I guess a single test is alright. Would [Christoph Gohlke's build be useful](http://www.lfd.uci.edu/~gohlke/pythonlibs/#numpy)?

Regarding py.test:
look here:
https://github.com/symengine/symengine.py/pull/4/files#diff-97e2e3c68308a6044377595c72c26cedR13

and here:
https://github.com/symengine/symengine.py/pull/4/files#diff-f92048e3a299f55719fea6172ce3b29dR41

certik commented:
@bjodah yes, we should use either the Gohlke's builds, or the binaries that numpy releases distribute upstream, or build our own using appveyor and store it in: https://github.com/symengine/dependencies/tree/binaries, see also https://github.com/symengine/dependencies/issues/4 about an idea how to do that easily.

We can test numpy with most builds if we use binary installers.

As to your git issues, definitely make sure you never force push into the PR. Then you can't lose your stuff. If you don't know what to do, post your branch somewhere, and we can have a look.

Ondrej

bjodah commented:
I changed the option to `WITH_NUMPY`, I removed numpy requirement from appveyor and refactored the tests so that nosetests just work in all situations.

This way dealing with pytest and numpy builds on windows could wait until future PRs.

All tests but one on travis seem to pass. During the installation of symengine:

```
-- Install configuration: "Release"
CMake Error at cmake_install.cmake:36 (FILE):
  file cannot create directory: /lib/cmake/symengine.  Maybe need
  administrative privileges.
```
any clues to why this would could be happening?

isuruf commented:
@bjodah, you have to update https://github.com/bjodah/symengine

bjodah commented:
@isuruf thanks! Updated to point real upstream since you merged the required PR.

certik commented on symengine/tests/test_var.py:
```diff
@@ -38,6 +38,13 @@ def test_var():
+def test_var_global_namespace():
     # see if var() really injects into global namespace
     raises(NameError, lambda: z1)
     _make_z1()
```
This is an important test that we need to keep. We'll have to figure out what is going on with py.test.

bjodah replied:
Yes, this is the very reason for why I implemented the workarounds to be able to use nosetests (contrary to what we agreed upon). Can we save the transition to `py.test` for a later PR?

certik replied:
I just don't like hackish solutions, and this feel very hackish, with unclear solution in sight. If you want to switch to py.test later, that's fine too, but then we should simply remove any py.test from this PR.

The reason I am saying that is that if you become busy after this PR is in, then we'll end up with a half baked transition between nosetests and py.test that somebody else will have to spend time to understand and untangle.

I am fine either way to use or not use py.test, but let's stick to one and follow through. I don't want to use both with different workarounds for each.

bjodah replied:
I see. I will remove the `py.test` related changes to this PR then (I'd much rather see that addressed in a separate PR). Should I leave things like `@py.test.mark.xfail(not HAVE_NUMPY, reason='array.array lacks "Zd"')` as a comment?

certik replied:
Yes, leave them commented out, that's a great idea. That way in a later PR, we can easily enable it once we switch to py.test.

certik commented on symengine/tests/test_lambdify.py:
```diff
@@ -0,0 +1,393 @@
+    A = l(inp, use_numpy=False)
+    check(A)
+
+# This test is currently failing due to missing bvisit method:
```
What kind of bvisit method is missing? Let's implement it while we are at it.

bjodah replied:
Sorry, this is a left over note. I will remove this note.

certik commented on symengine/tests/test_lambdify.py:
```diff
@@ -0,0 +1,393 @@
+    l, check = _get_2_to_2by2_list()
+    inp = [13, 17]
+    A = l(inp, use_numpy=False)
+    check(A, inp)
```
Why does this fail?

bjodah replied:
Good question, can't remember at the moment. I will look into it and document it properly.

certik commented on symengine/tests/test_lambdify.py:
```diff
@@ -0,0 +1,393 @@
+@pytest.mark.skipif(not HAVE_NUMPY, reason='requires numpy')
+def test_2_to_2by2_numpy():
+    if not HAVE_NUMPY:  # nosetests work-around
+        return
```
We agreed to just use py.test, so we don't need to add any nosetests workarounds any more.

bjodah replied:
Se my reply to your comment above.

certik commented on cmake/cython_test.pyx:
```diff
@@ -83,3 +83,9 @@ cdef extern from "<symengine/add.h>" namespace "SymEngine":
+# Currently not used:
+# Test that cpdef void is supported (https://github.com/cython/cython/commit/7df9d654b3e6c3d5812a95904872d35a3c94ef77)
+# cpdef void func(double[::1] inp):
+#     pass
```
Let's remove it if we don't need it.

bjodah replied:
Ok, will do.

certik commented on bin/test_travis.sh:
```diff
@@ -11,6 +11,7 @@ python setup.py install --symengine-dir=$our_install_dir
 # Test python wrappers
 if [[ "${WITH_SAGE}" != "yes" ]]; then
     nosetests -v
+    py.test -v symengine/tests/test_*.py  # pytest gets confused if we install first
```
let's remove nosetests, and just use py.test, as we agreed.

I think you can use `$PYTHON_SOURCE_DIR` as below for the path?

certik replied:
See also https://github.com/symengine/symengine.py/blob/686bff8bbbb4945d0dc826b6ff7485bf164b46e2/symengine/__init__.py#L16, which already uses py.test.

bjodah replied:
See my reply to your above comment. Regarding `$PYTHON_SOURCE_DIR`, do you mean like this:
`py.test -v $PYTHON_SOURCE_DIR/symengine/tests/test_*.py`

certik commented:
Generally it looks good, but there are still quite a few technical things to resolve.

Since this PR depends on py.test, I propose we do the nosetests -> py.test switch in a separate PR, since we need to make sure all tests pass, and that the switch didn't break anything. Currently the way the switch is implemented in this PR, it breaks a few things. So let's iron our the py.test issues in a new PR, and once that one is merged, you don't need to worry about py.test anymore in this PR.

bjodah commented:
@certik sorry, should have read your final comment first. So I'll leave the `py.test`/`nosetests` issue be for now and address the other comments.

bjodah commented:
Closed in favor of #11
